File: inventur/inventur.go
package inventur

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
)

var (
	RootPath = filepath.Join("inventur", "Data")
)

type AllFile struct {
	Artikelnummer string `json:"Artikelnummer"`
	Suchbegriff   string `json:"Suchbegriff"`
	Anzahl        int    `json:"Anzahl"`
	Team          string `json:"Team"`
}

func GetYears(w http.ResponseWriter, r *http.Request) {
	ex, err := os.Executable()
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}
	exPath := filepath.Dir(ex)

	folder, err := os.ReadDir(filepath.Join(exPath, RootPath))
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}
	var folderNames []string
	for _, e := range folder {
		folderNames = append(folderNames, e.Name())
	}
	json.NewEncoder(w).Encode(folderNames)
}

func GetAllEntries(w http.ResponseWriter, r *http.Request) {
	year := r.FormValue("Year")
	ex, err := os.Executable()
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}
	exPath := filepath.Dir(ex)

	file, err := os.ReadFile(fmt.Sprintf("%s/_ALL.json", filepath.Join(exPath, RootPath, year)))
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	var res []AllFile
	err = json.Unmarshal([]byte(file), &res)
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	json.NewEncoder(w).Encode(res)
}

type TeamFile struct {
	Team        int    `json:"Team"`
	Mitarbeiter string `json:"Mitarbeiter"`
	Farbe       string `json:"Farbe"`
	Ort         string `json:"Ort"`
}

func GetTeams(w http.ResponseWriter, r *http.Request) {
	year := r.FormValue("Year")
	ex, err := os.Executable()
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}
	exPath := filepath.Dir(ex)

	file, err := os.ReadFile(fmt.Sprintf("%s/_TEAMS.json", filepath.Join(exPath, RootPath, year)))

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	var res []TeamFile
	err = json.Unmarshal([]byte(file), &res)
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	json.NewEncoder(w).Encode(res)
}

type Entry struct {
	Artikelnummer string `json:"Artikelnummer"`
	Suchbegriff   string `json:"Suchbegriff"`
	Anzahl        int    `json:"Anzahl"`
}

func GetEntry(w http.ResponseWriter, r *http.Request) {
	year := r.FormValue("Year")
	team := r.FormValue("Team")
	ex, err := os.Executable()
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}
	exPath := filepath.Dir(ex)

	file, err := os.ReadFile(fmt.Sprintf("%s/%s.json", filepath.Join(exPath, RootPath, year), team))

	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	var res []Entry
	err = json.Unmarshal([]byte(file), &res)
	if err != nil {
		fehler := err.Error()
		json.NewEncoder(w).Encode(map[string]string{"error": fehler})
		return
	}

	json.NewEncoder(w).Encode(res)
}

File: src/db/Inventur.ts
import { z } from "zod";
import { client, config } from "./config";

const getYears = async (): Promise<string[]> => {
  const res = await client.get<string[]>("/Inventur/Years", config);
  return res.data;
};

const GetEntriesProps = z.object({
  Year: z.string(),
});
type GetEntriesProps = z.infer<typeof GetEntriesProps>;

const AllFile = z.object({
  Artikelnummer: z.string(),
  Suchbegriff: z.string(),
  Anzahl: z.number().int(),
  Team: z.string(),
});
type AllFile = z.infer<typeof AllFile>;

const getEntries = async (props: GetEntriesProps): Promise<AllFile[]> => {
  const form = new FormData();
  form.set("Year", props.Year);
  const res = await client.post<AllFile[]>("/Inventur/All", form, config);
  return res.data;
};

const TeamFile = z.object({
  Team: z.number().int(),
  Mitarbeiter: z.string(),
  Farbe: z.string(),
  Ort: z.string(),
});
type TeamFile = z.infer<typeof TeamFile>;

const GetTeamsProps = z.object({
  Year: z.string(),
});
type GetTeamsProps = z.infer<typeof GetTeamsProps>;

const getTeams = async (props: GetTeamsProps): Promise<TeamFile[]> => {
  const form = new FormData();
  form.set("Year", props.Year);
  const res = await client.post<TeamFile[]>("/Inventur/Teams", form, config);
  return res.data;
};

const Entry = z.object({
  Artikelnummer: z.string(),
  Suchbegriff: z.string(),
  Anzahl: z.number().int(),
});
type Entry = z.infer<typeof Entry>;

const GetEntryProps = z.object({
  Year: z.string(),
  Team: z.string(),
});
type GetEntryProps = z.infer<typeof GetEntryProps>;

const getEntry = async (props: GetEntryProps): Promise<Entry[]> => {
  const form = new FormData();
  form.set("Year", props.Year);
  form.set("Team", props.Team);
  const res = await client.post<Entry[]>("/Inventur/Entry", form, config);
  return res.data;
};

export {
  getYears,
  getEntries,
  getTeams,
  getEntry,
  GetEntriesProps,
  GetTeamsProps,
  GetEntryProps,
  Entry,
  AllFile,
  TeamFile,
};
